### ISSUE TITLE
grpc+verbs not being used

### ISSUE THREAD

**Issue by limjingrong** [tensorflow/tensorflow#34808]
<em>Please make sure that this is a build/installation issue. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:build_template</em>

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 16.04 LTS
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N.A
- TensorFlow installed from (source or binary): source
- TensorFlow version: 1.15
- Python version: 3.5.2, 3.6.8 (different machines)
- Installed using virtualenv? pip? conda?: pip
- Bazel version (if compiling from source): 0.26.1
- GCC/Compiler version (if compiling from source): GCC 7.4
- CUDA/cuDNN version: NA
- GPU model and memory: NA

**Describe the problem**
Running code with tf.estimator.RunConfig(...., protocol="grpc+verbs") does not throw any error, but does not appear to use RDMA (logs do not show any mention of RDMA). Running the code without this protocol (i.e just on gRPC) appears to work fine, which leads me to think that it's a build / installation issue. We also observe no speed-up by using RDMA, which suggests that RDMA is not being used. We are using a 1x parameter server, 1x chief, 1x evaluator, 15x worker setup (Parameter server, chief, and evaluator are on the same node and share the same filesystem; the workers are on separate machines and do not share the same file system).

**Provide the exact sequence of commands / steps that you executed before running into the problem**
I followed the guide [here](https://www.tensorflow.org/install/source) to compile from source. However, running the "./configure" step does **not** give an option to install with verbs support. Hence, I built Tensorflow with the following command:
`bazel build --config=verbs //tensorflow/tools/pip_package:build_pip_package`. Everything else followed the given guide.

**Any other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

Sample parameter server logs:
`2019-12-03 13:26:22.885842: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX
WARNING:tensorflow:
The TensorFlow contrib module will not be included in TensorFlow 2.0.
For more information, please see:
  * https://github.com/tensorflow/community/blob/master/rfcs/20180907-contrib-sunset.md
  * https://github.com/tensorflow/addons
  * https://github.com/tensorflow/io (for I/O related ops)
If you depend on functionality not listed there, please file an issue.

WARNING:tensorflow:
The TensorFlow contrib module will not be included in TensorFlow 2.0.
For more information, please see:
  * https://github.com/tensorflow/community/blob/master/rfcs/20180907-contrib-sunset.md
  * https://github.com/tensorflow/addons
  * https://github.com/tensorflow/io (for I/O related ops)
If you depend on functionality not listed there, please file an issue.

WARNING:tensorflow:`eval_strategy` is not passed in. No distribution strategy will be used for evaluation.
WARNING:tensorflow:`eval_strategy` is not passed in. No distribution strategy will be used for evaluation.
2019-12-03 13:26:22.899183: I tensorflow/core/distributed_runtime/rpc/grpc_channel.cc:258] Initialize GrpcChannelCache for job chief -> {0 -> 192.168.9.20:1235}
2019-12-03 13:26:22.899214: I tensorflow/core/distributed_runtime/rpc/grpc_channel.cc:258] Initialize GrpcChannelCache for job ps -> {0 -> localhost:1236}
2019-12-03 13:26:22.899232: I tensorflow/core/distributed_runtime/rpc/grpc_channel.cc:258] Initialize GrpcChannelCache for job worker -> {0 -> 192.168.9.16:1234, 1 -> 192.168.9.17:1234, 2 -> 192.168.9.18:1234, 3 -> 192.168.9.23:1234, 4 -> 192.168.9.24:1234, 5 -> 192.168.9.25:1234, 6 -> 192.168.9.26:1234, 7 -> 192.168.9.27:1234, 8 -> 192.168.9.28:1234, 9 -> 192.168.9.29:1234, 10 -> 192.168.9.30:1234, 11 -> 192.168.9.31:1234, 12 -> 192.168.9.32:1234, 13 -> 192.168.9.105:1234, 14 -> 192.168.9.106:1234}
2019-12-03 13:26:22.903725: I tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc:365] Started server with target: grpc://localhost:1236
2019-12-03 13:26:22.903796: I tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc:369] Server already started (target: grpc://localhost:1236)
2019-12-03 13:26:28.930441: I tensorflow/core/distributed_runtime/worker.cc:203] Cancellation requested for RunGraph.
2019-12-03 13:26:29.073805: I tensorflow/core/distributed_runtime/worker.cc:203] Cancellation requested for RunGraph.
2019-12-03 13:26:29.308595: I tensorflow/core/distributed_runtime/worker.cc:203] Cancellation requested for RunGraph.
2019-12-03 13:26:29.609207: I tensorflow/core/distributed_runtime/worker.cc:203] Cancellation requested for RunGraph.
2019-12-03 13:26:29.994561: I tensorflow/core/distributed_runtime/worker.cc:203] Cancellation requested for RunGraph.
2019-12-03 13:26:30.430739: I tensorflow/core/distributed_runtime/worker.cc:203] Cancellation requested for RunGraph.
2019-12-03 13:26:30.971334: I tensorflow/core/distributed_runtime/worker.cc:203] Cancellation requested for RunGraph.
2019-12-03 13:26:59.046948: I tensorflow/core/distributed_runtime/worker.cc:203] Cancellation requested for RunGraph.
2019-12-03 13:26:59.278155: I tensorflow/core/distributed_runtime/worker.cc:203] Cancellation requested for RunGraph.`

**Comment by byronyi** [comment]
It seems your RDMA configuration is not in effective. Would you mind to check again with official TensorFlow package (without verbs support) and see if there is any error reported?

**Comment by limjingrong** [comment]
Hmm. @byronyi do you have a basic RDMA test code that I could use? I want to rule out the possibility that it's a quirk with the code I'm using--it seems like estimators are pretty new to distributed training.

**Comment by byronyi** [comment]
@limjingrong Back in the old days we use tensorflow/benchmarks, as shown in one of my old (and stale) issues:

https://github.com/tensorflow/tensorflow/issues/23606

Just change the server_protocol parameter from grpc to grpc+verbs to test it out.

**Comment by limjingrong** [comment]
@byronyi Thanks for the suggestion! I ran the benchmarks from the branch cnn_tf_v1.12_compatible with 
`python3 benchmarks/scripts/tf_cnn_benchmarks/tf_cnn_benchmarks.py \
  --ps_hosts=192.168.9.20:1236 \
  --worker_hosts=192.168.9.18:1234 \
  --job_name=worker \
  --task_index=0 \
  --server_protocol=grpc+verbs \
  --variable_update=parameter_server \
  --local_parameter_device=cpu \
  --model=resnet50 \
  --use_fp16 \
  --batch_size=256 \
  --device=cpu
  --data_format=NHWC
  --num_warmup_batches=1`

(I ran into another unrelated problem, where the worker hangs while warming up, but that may be related to me using CPU only)

The parameter server in this set up outputs the line
`2019-12-06 14:58:39.442051: I tensorflow/contrib/verbs/rdma_mgr.cc:130] Connected to remote node /job:worker/replica:0/task:0`, which verifies that RDMA is part of the build. I'm now wondering if grpc+verbs isn't fully supported by the new TF estimator framework--I'm using _tf.distribute.experimental.ParameterServerStrategy_

Our code is as follows
'
    os.environ["TF_CONFIG"]  = json.dumps({
           "cluster": {
               "worker": ["192.168.9.18:1234"],
               "chief": ["192.168.9.20:1235"],
               "ps": ["192.168.9.20:1236"],
               "evaluator": ["192.168.9.20:1237"]
       },
          "task": {"type": "chief", "index": 0}
    })

    BUFFER_SIZE = 10000
    BATCH_SIZE = 32
    EPOCHS = 10

    def input_fn(mode, input_context=None):
        datasets, info = tfds.load(name="mnist",
                                    with_info=True,
                                    as_supervised=True)
        mnist_dataset = (datasets["train"] if mode == tf.estimator.ModeKeys.TRAIN else
                       datasets["test"])

        def scale(image, label):
            image = tf.cast(image, tf.float32)
            image /= 255
            return image, label

        if input_context:
            mnist_dataset = mnist_dataset.shard(input_context.num_input_pipelines,
                                            input_context.input_pipeline_id)
        return mnist_dataset.map(scale).cache().repeat(EPOCHS).shuffle(BUFFER_SIZE).batch(BATCH_SIZE)

    datasets, info = tfds.load(name="mnist",
                                with_info=True,
                                as_supervised=True)
    mnist_dataset = (datasets["train"])

    def scale(image, label):
        image = tf.cast(image, tf.float32)
        image /= 255
        return image, label

    mnist_dataset.map(scale).cache().repeat(EPOCHS).shuffle(BUFFER_SIZE).batch(BATCH_SIZE)
    LEARNING_RATE = 1e-1

    def model_fn(features, labels, mode):
        training = mode == tf.estimator.ModeKeys.TRAIN
        model = tf.keras.Sequential([
            tf.keras.layers.Flatten(),
            tf.keras.layers.Dense(10)
      ])
        logits = model(features, training)

        if mode == tf.estimator.ModeKeys.PREDICT:
            predictions = {'logits': logits}
            return tf.estimator.EstimatorSpec(labels=labels, predictions=predictions)

        optimizer = tf.train.GradientDescentOptimizer(learning_rate=LEARNING_RATE)
        loss = tf.keras.losses.SparseCategoricalCrossentropy(
          reduction="none",
          from_logits=True)(labels, logits)
        loss = tf.reduce_sum(loss) * (1. / BATCH_SIZE)
        if mode == tf.estimator.ModeKeys.EVAL:
            metrics_dict = {
                'accuracy': tf.metrics.accuracy(labels, tf.argmax(logits, axis=-1))
            }
            return tf.estimator.EstimatorSpec(mode, loss=loss, eval_metric_ops=metrics_dict)

        train_op = optimizer.minimize(
            loss, global_step=tf.train.get_or_create_global_step())

        return tf.estimator.EstimatorSpec(
          mode=mode,
          loss=loss,
          train_op=train_op)

    strategy = tf.distribute.experimental.ParameterServerStrategy()
    run_config = tf.estimator.RunConfig(
        experimental_distribute=tf.contrib.distribute.DistributeConfig(
           train_distribute=strategy,
        ),
        protocol="grpc+verbs"
    )
    estimator = tf.estimator.Estimator(config=run_config, model_fn=model_fn, model_dir="/home/rdma_for_ml/model_dir")

    start = time.clock()
    tf.estimator.train_and_evaluate(
        estimator,
        train_spec=tf.estimator.TrainSpec(input_fn=input_fn),
        eval_spec=tf.estimator.EvalSpec(input_fn=input_fn))
    print(time.clock()-start)
`

**Comment by tensorflowbutler** [comment]
Hi There,

 We are checking to see if you still need help on this, as you are using an older version of tensorflow which is officially considered end of life . We recommend that you upgrade to the latest 2.x version and let us know if the issue still persists in newer versions. Please open a new issue for any help you need against 2.x, and we will get you the right help. 

 This issue will be closed automatically 7 days from now. If you still need help with this issue, please provide us with more information.
